programs/clearing_house/src/lib.rs

```
@@ -747,6 +749,210 @@ pub mod clearing_house {
+
+        // let mut oracle_map = OracleMap::load(remaining_accounts_iter, clock.slot)?;
+        // let bank_map = BankMap::load(
+        //     &get_writable_banks(QUOTE_ASSET_BANK_INDEX),
+        //     remaining_accounts_iter,
+        // )?;
+    
+        let market_map = MarketMap::load(
+            &get_writable_markets(market_index),
+            &get_market_oracles(market_index, &ctx.accounts.oracle),
+            remaining_accounts_iter,
+        )?;
+        let mut market = market_map.get_ref_mut(&market_index)?;
+
+        // TODO: margin requirements
+        // only 1x leverage fn 
+
+        let position_index = get_position_index(&user.positions, market_index)
+            .or_else(|_| add_new_position(&mut user.positions, market_index))?;
+        let lp_position = &mut user.positions[position_index];
+
+        // distribute lp tokens to lp
+        let user_lp_tokens = quote_asset_amount * AMM_TO_QUOTE_PRECISION_RATIO
+            / 2
+            / (market.amm.peg_multiplier / PEG_PRECISION); // todo: change from peg to mark price?
+
+        // update market state
+        let total_lp_tokens = market.amm.sqrt_k;
+        let reserve_scale = ((total_lp_tokens + user_lp_tokens) * AMM_RESERVE_PRECISION) / total_lp_tokens;
+
+        market.amm.base_asset_reserve = market.amm.base_asset_reserve
+            .checked_mul(reserve_scale)
+            .ok_or_else(math_error!())?
+            .checked_div(AMM_RESERVE_PRECISION)
+            .ok_or_else(math_error!())?;
+
+        market.amm.quote_asset_reserve = market.amm.quote_asset_reserve
+            .checked_mul(reserve_scale)
+            .ok_or_else(math_error!())?
+            .checked_div(AMM_RESERVE_PRECISION)
+            .ok_or_else(math_error!())?;
+
+        market.amm.sqrt_k = market.amm.sqrt_k
+            .checked_add(user_lp_tokens)
+            .ok_or_else(math_error!())?;
+
+        // update lp position
+        lp_position.lp_tokens = user_lp_tokens;
```

do you need to throw an error if the user already has a normal position?